### DISCUSSION TITLE
Type checking and abstracting table classes

### DISCUSSION THREAD

**Discussion by supern8ent** [sqlalchemy/sqlalchemy#10318] [Usage Questions] [answered]
I have some tables - defined in the orm style - that have the same structure, but I prefer to keep them as separate tables. There is some functionality that operates on these tables, so to follow DRY I've abstracted the code so it accepts the orm class as an argument. Operationally it's fine, but type checking is not as straightforward as I hoped and would like to ask for advice.

Here's a dumbed down version of my issue:
```python
# sqa_problem.py

from typing import Type, Union
from typing_extensions import Annotated
from sqlalchemy import create_engine, select
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, Session

intpk = Annotated[int, mapped_column(primary_key=True)]


class Base(DeclarativeBase):
    pass


class Table1(Base):
    __tablename__ = "table_1"
    id: Mapped[intpk]
    name: Mapped[str]


class Table2(Base):
    __tablename__ = "table_2"
    id: Mapped[intpk]
    name: Mapped[str]


def print_names(session: Session, TableClass: Union[Type[Table1], Type[Table2]]) -> None:
    rows = session.scalars(select(TableClass)).all()
    print(", ".join([r.name for r in rows]))


engine = create_engine("postgresql://postgres@localhost:5432/testdb")
Base.metadata.create_all(engine)

with Session(engine) as session, session.begin():
    for n in ["bill", "amy", "sekhar"]:
        row_1 = Table1(name=n)
        session.add(row_1)
        row_2 = Table2(name=n[::-1])
        session.add(row_2)

with Session(engine) as session:
    for TableClass in [Table1, Table2]:
        print_names(session, TableClass)
```

```bash
$ mypy sqa_problem.py
sqa_problem.py:27: error: "Base" has no attribute "name"  [attr-defined]
sqa_problem.py:42: error: Argument 2 to "print_names" has incompatible type "type[Base]"; expected "Union[type[Table1], type[Table2]]"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```
Versions:
- python 3.9.16
- sqlalchemy 2.0.20
- mypy 1.5.1

I found a couple of workarounds:
1. In `print_names` hint `TableClass` as `Any`
1. In `print_names` hint `TableClass` as `Type(Base)` and `cast` the rows to type `Table1`
1. In `print_names` hint `TableClass` as `Type(Base)` and add `# type: ignore` on the `print` line

The code as is feels like it should ideally work, and I wonder if Sqlalchemy 2 is working toward enabling that syntax. I'm also surprised that mypy assumes that type `Base` is being passed to `print_names` here, as opposed to `Table1` and `Table2`.

**Comment by CaselIT** [comment]
Hi

> The code as is feels like it should ideally work, and I wonder if Sqlalchemy 2 is working toward enabling that syntax. 

I'm not sure what sqlalchemy could do to "enable" that syntax. This behaviour is defined by how python typing works, and I don't see any way of changing it in the sqlalchemy library 

> I'm also surprised that mypy assumes that type Base is being passed to print_names here, as opposed to Table1 and Table2.

The common types of the two classes is base, so that's what the list gets typed to. Maybe try using a tuple to check if things improve

**Reply by supern8ent** [reply]
> This behaviour is defined by how python typing works, and I don't see any way of changing it in the sqlalchemy library

I tried some other things and I see what you mean.

> Maybe try using a tuple to check if things improve
I tried the following:
```python
with Session(engine) as session:
    my_models: tuple[Type[Table1], Type[Table2]] = (Table1, Table2)
    for TableClass in my_models:
        print_names(session, TableClass)
```
mypy still assumed type `Base` was being passed to `print_names`.

**Comment by zzzeek** [comment, marked as answer]
`Base` is the common point between `Table1` and `Table2` so that's likely what mypy is doing here.    What might be best here is use a Protocol as we had a user doing in #10287.    

```py
# sqa_problem.py

from typing import cast
from typing import Protocol
from typing import Type

from typing_extensions import Annotated

from sqlalchemy import create_engine
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session

intpk = Annotated[int, mapped_column(primary_key=True)]


class Base(DeclarativeBase):
    pass


class HasName(Protocol):
    name: Mapped[str]

class Table1(Base):
    __tablename__ = "table_1"
    id: Mapped[intpk]
    name: Mapped[str]


class Table2(Base):
    __tablename__ = "table_2"
    id: Mapped[intpk]
    name: Mapped[str]


def print_names(session: Session, TableClass: Type[HasName]) -> None:
    rows = session.scalars(select(TableClass)).all()
    print(", ".join([r.name for r in rows]))


engine = create_engine("postgresql://postgres@localhost:5432/testdb")
Base.metadata.create_all(engine)

with Session(engine) as session, session.begin():
    for n in ["bill", "amy", "sekhar"]:
        row_1 = Table1(name=n)
        session.add(row_1)
        row_2 = Table2(name=n[::-1])
        session.add(row_2)

with Session(engine) as session:

    for TableClass in [Table1, Table2]:
        print_names(session, cast(Type[HasName], TableClass))
```


I tried using a real mixin also but mypy did not have any better time with the `print_names()` at the bottom.    Python typing is currently much less flexible passing around types of classes unfortunately.

**Comment by supern8ent** [comment]
Thank you @CaselIT and @zzzeek for helping me understand the issue better and especially @zzzeek for the solution using Protocols.

@CaselIT and @zzzeek, you both alluded to limitations in python typing. I came up with an relevant example.

Similar problem, but passing instances instead of classes:
```python
# sqa_problem_non_db.py

from typing import Union

class Table1:
    def __init__(self, name: str):
        self.id: int = 0
        self.name: str = name

class Table2:
    def __init__(self, name: str):
        self.id: int = 0
        self.name: str = name

def print_name(tinst: Union[Table1, Table2]) -> None:
    print(tinst.name)

table1_instance = Table1("bill")
table2_instance = Table2("amy")

for table in [table1_instance, table2_instance]:
    print_name(table)
```
```
$ mypy sqa_problem_non_db.py
sqa_problem_non_db.py:24: error: Argument 1 to "print_name" has incompatible type "object"; expected "Union[Table1, Table2]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

That error can be resolved by typing the list:
```python
table_instances: list[Union[Table1, Table2]] = [table1_instance, table2_instance]
for table in table_instances:
    print_name(table)
```

The analogous solution to my original issue would be:
```python
with Session(engine) as session:
    my_models: list[Union[Type[Table1], Type[Table2]]] = [Table1, Table2]
    for m in my_models:
        print_names(session, m)
```
But mypy doesn't act in an analogous fashion. It still assumes `Base`:
```
mypy sqa_problem_soln5.py
sqa_problem_soln5.py:27: error: "Base" has no attribute "name"  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
```
